### PR TITLE
pidgin: fix build on darwin, add mainProgram

### PR DESCRIPTION
**Pull request by lucasew**
## Description of changes
I don't have anything darwin here and I am the only maintainer of pidgin.

ZHF #309482 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/networking/instant-messengers/pidgin/default.nix [hunk 1]
```diff
@@ -141,6 +141,7 @@ let
       homepage = "https://pidgin.im/";
       license = lib.licenses.gpl2Plus;
       platforms = lib.platforms.unix;
+      broken = stdenv.isDarwin;
```
Comment (AndersonTorres) [inline pkgs/applications/networking/instant-messengers/pidgin/default.nix:144]: where is mainProgram?

**Review by ghost — COMMENTED** [review]
the diff below fixes the clang build. (tested on x64 darwin). can probably also test using `pkgsLLVM.pidgin` on linux, but that needs to build a bunch of stuff so i didn't try.

<details><summary>ignore clang error.diff</summary>

```diff
diff --git a/pkgs/applications/networking/instant-messengers/pidgin/default.nix b/pkgs/applications/networking/instant-messengers/pidgin/default.nix
index 211c5b49bf9a..4274935caf5f 100644
--- a/pkgs/applications/networking/instant-messengers/pidgin/default.nix
+++ b/pkgs/applications/networking/instant-messengers/pidgin/default.nix
@@ -107,7 +107,8 @@ let
     ]
     ++ lib.optionals withCyrus_sasl [ "--enable-cyrus-sasl=yes" ]
     ++ lib.optionals withGnutls [ "--enable-gnutls=yes" "--enable-nss=no" ]
-    ++ lib.optionals stdenv.isDarwin [ "--disable-gtkspell" "--disable-vv" ];
+    ++ lib.optionals stdenv.isDarwin [ "--disable-gtkspell" "--disable-vv" ]
+    ++ lib.optionals stdenv.cc.isClang [ "CFLAGS=-Wno-error=int-conversion" ];
 
     enableParallelBuilding = true;
 
```
</details>

File: pkgs/applications/networking/instant-messengers/pidgin/default.nix (hunk 1 above)
Comment (lucasew) [reply pkgs/applications/networking/instant-messengers/pidgin/default.nix]: Fixed

**Comment by lucasew** [comment]
> the diff below fixes the clang build. (tested on x64 darwin). can probably also test using `pkgsLLVM.pidgin` on linux, but that needs to build a bunch of stuff so i didn't try.
> 
> ignore clang error.diff

I applied that change. Can you run nixpkgs-review on this PR?

**Comment by lucasew** [comment]
@ofborg build pidgin

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 310923` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 310923` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>pidginPackages.tdlib-purple</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>telepathy-haze</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>pidgin (pidginPackages.pidgin)</li>
    <li>pidginPackages.purple-matrix</li>
  </ul>
</details>

**Review by ghost — APPROVED** [review]
Result of `nixpkgs-review pr 310923` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>pidginPackages.tdlib-purple</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>telepathy-haze</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>pidgin (pidginPackages.pidgin)</li>
    <li>pidginPackages.purple-matrix</li>
  </ul>
</details>

`telepathy-haze` fails due to unrelated compile error.

**Pull request merged**